bindings/src/test/java/io/pravega/storage/extendeds3/S3Mock.java

```diff
@@ -0,0 +1,315 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.storage.extendeds3;
+
+import com.emc.object.Range;
+import com.emc.object.s3.S3Exception;
+import com.emc.object.s3.S3ObjectMetadata;
+import com.emc.object.s3.bean.AccessControlList;
+import com.emc.object.s3.bean.CompleteMultipartUploadResult;
+import com.emc.object.s3.bean.CopyPartResult;
+import com.emc.object.s3.bean.DeleteObjectsResult;
+import com.emc.object.s3.bean.GetObjectResult;
+import com.emc.object.s3.bean.ListObjectsResult;
+import com.emc.object.s3.bean.ObjectKey;
+import com.emc.object.s3.bean.PutObjectResult;
+import com.emc.object.s3.bean.S3Object;
+import com.emc.object.s3.request.AbortMultipartUploadRequest;
+import com.emc.object.s3.request.CompleteMultipartUploadRequest;
+import com.emc.object.s3.request.CopyPartRequest;
+import com.emc.object.s3.request.DeleteObjectsRequest;
+import com.emc.object.s3.request.PutObjectRequest;
+import com.emc.object.s3.request.SetObjectAclRequest;
+import io.pravega.common.io.StreamHelpers;
+import io.pravega.common.util.BufferView;
+import io.pravega.common.util.ByteArraySegment;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+import lombok.AllArgsConstructor;
+import lombok.val;
+
+/**
+ * In-memory mock for S3.
+ */
+public class S3Mock {
+    //region Private
+
+    @GuardedBy("objects")
+    private final Map<String, Map<Integer, CopyPartRequest>> multipartUploads;
+    @GuardedBy("objects")
+    private final Map<String, ObjectData> objects;
+
+    //endregion
+
+    //region Constructor
+
+    public S3Mock() {
+        this.objects = new HashMap<>();
+        this.multipartUploads = new HashMap<>();
+    }
+
+    //endregion
+
+    //region Mock Implementation
+
+    private String getObjectName(String bucketName, String key) {
+        return String.format("%s-%s", bucketName, key);
+    }
+
+    public PutObjectResult putObject(PutObjectRequest request) {
+        if (request.getObjectMetadata() != null) {
+            request.setObjectMetadata(null);
+        }
+        String objectName = getObjectName(request.getBucketName(), request.getKey());
+        synchronized (this.objects) {
+            if (this.objects.containsKey(objectName)) {
+                throw new S3Exception(String.format("Object '%s/%s' exists.", request.getBucketName(), request.getKey()), 0, "PreconditionFailed", "");
```

Comment:
HTTP code for precondition failed is 412. Let's use constant from org.apache.http.HttpStatus

Reply:
Would love to do so, but that package is not defined as a dependency of this project (segmentstore:storage(test)).

Reply:
Thanks for the pointer. I was looking at the wrong thing. I updated all response codes in this class.